### PR TITLE
New docs for Require Registration

### PR DESCRIPTION
**Pull request by brettwp**
Adds documentation for "Require Registration" toggle on `/core-concepts/applications` and API field on `/apis/applications/` as well as the various include `.adoc` and `.json` files.

### CODE REVIEW

**Comment by robotdan** [comment]
Enable grants in response is missing `urn:ietf:params:oauth:grant-type:device_code`, see `GrantType`

File: site/docs/v1/tech/apis/_application-oauth-configuration-response-body.adoc [hunk 1]
```diff
@@ -46,8 +46,11 @@ The logout URL for the Application. FusionAuth will redirect to this URL after t
 [field]#application.oauthConfiguration.requireClientAuthentication# [type]#[Boolean]# [since]#Available since 1.3.0#::
 Determines if the OAuth 2.0 Token endpoint requires client authentication. If this is enabled, the client must provide client credentials when using the Token endpoint. The `client_id` and `client_secret` may be provided using a Basic Authorization HTTP header, or by sending these parameters in the request body using POST data.
 
+[field]#application.oauthConfiguration.requireRegistration# [type]#[Boolean]# [optional]#Optional# [default]#Defaults to `false`# [since]#Available since 1.28.0#::
```
Comment (robotdan) [inline site/docs/v1/tech/apis/_application-oauth-configuration-response-body.adoc:49]: > This configuration does not affect any other grant, and does not affect the API usage.

File: site/docs/v1/tech/apis/_application-request-body.adoc [hunk 2]
```diff
@@ -197,6 +197,9 @@ Determines if the OAuth 2.0 Token endpoint requires client authentication. If th
 [deprecated]#Deprecated in version 1.28.0#
 In version 1.28.0 and beyond, client authentication can be managed via [field]#application.oauthConfiguration.clientAuthenticationPolicy#.
 
+[field]#application.oauthConfiguration.requireRegistration# [type]#[Boolean]# [optional]#Optional# [default]#Defaults to `false`# [since]#Available since 1.28.0#::
```
Comment (robotdan) [inline site/docs/v1/tech/apis/_application-request-body.adoc:200]: > This configuration does not affect any other grant, and does not affect the API usage.

File: site/docs/v1/tech/apis/_application-oauth-configuration-response-body.adoc (hunk 1 above)
Comment (robotdan) [reply site/docs/v1/tech/apis/_application-oauth-configuration-response-body.adoc:49]: The response objects don't need the default or optional flags.

File: site/docs/v1/tech/apis/_application-response-body-base.adoc [hunk 3]
```diff
@@ -179,6 +179,9 @@ Determines if the OAuth 2.0 Token endpoint requires client authentication. If th
 [deprecated]#Deprecated in version 1.28.0#
 In version 1.28.0 and beyond, client authentication can be managed via [field]#application.oauthConfiguration.clientAuthenticationPolicy#.
 
+[field]#application.oauthConfiguration.requireRegistration# [type]#[Boolean]# [since]#Available since 1.28.0#::
+Determines if the user will be required to be registered, or complete registration before redirecting to the configured callback in the authorization code grant or the implicit grant. This configuration does not currently apply to any other grant.
```
Comment (robotdan) [inline site/docs/v1/tech/apis/_application-response-body-base.adoc:183]: > This configuration does not affect any other grant, and does not affect the API usage.

File: site/docs/v1/tech/core-concepts/applications.adoc [hunk 4]
```diff
@@ -152,6 +151,9 @@ The enabled OAuth2 grants. If a grant is not enabled and a client requests this
 +
 See The link:/docs/v1/tech/oauth/[OAuth 2.0 & OpenID Connect Overview] for additional information on each of these grants.
 
+[field]#Require Registration# [optional]#Optional#::
```
Comment (robotdan) [inline site/docs/v1/tech/core-concepts/applications.adoc:154]: We could document the default value

File: site/docs/v1/tech/core-concepts/applications.adoc [hunk 5]
```diff
@@ -152,6 +151,9 @@ The enabled OAuth2 grants. If a grant is not enabled and a client requests this
...
+When enabled the user will be required to be registered, or complete registration before redirecting to the configured callback in the authorization code grant or the implicit grant. This configuration does not currently apply to any other grant.
```
Comment (robotdan) [inline site/docs/v1/tech/core-concepts/applications.adoc:155]: > This configuration does not affect any other grant, and does not affect the API usage.

**Comment by mooreds** [comment]
@brettwp it would also be good to update https://fusionauth.io/docs/v1/tech/core-concepts/authentication-authorization/ to reflect this option. I'd add something like:

```
If you are utilizing the JWT to authorize a user to your application, you must do more than just ensure the JWT has a valid signature and is not expired. You must also ensure the JWT has provided adequate claims to the user’s authorization. (existing)

If you enable "Require Registration" on an application, the JWT won't be provided until the user is registered for this application.
```

File: site/docs/v1/tech/apis/_application-oauth-configuration-response-body.adoc (hunk 1 above)
Comment (brettwp) [reply site/docs/v1/tech/apis/_application-oauth-configuration-response-body.adoc:49]: Good catch.  I gotta get better about copy-n-paste.

**Comment by brettwp** [comment]
> Enable grants in response is missing `urn:ietf:params:oauth:grant-type:device_code`, see `GrantType`, this isn't related to your change, but while you're in here, can sync it up.

@robotdan This PR changes 3 `response.json` files and they all have `"enabledGrants": [...]` arrays.  I'm not sure which ones to add `urn:ietf:params:oauth:grant-type:device_code` to or if that is even the correct place.  Could you elaborate?

File: site/docs/v1/tech/core-concepts/authentication-authorization.adoc [hunk 6]
```diff
@@ -45,19 +45,21 @@ include::docs/src/json/login/jwt.json[]
...
+If you enable "Require Registration" on an application, the JWT won't be provided until the user is registered for this application.
+
 *Checking the JWT signature and expiration are only a part of the story.*
 
 The `aud` claim identifies the context of the request, in other words who is this JWT for: a Payroll application, a mobile application, etc. The presence of the `applicationId` and `roles` claims identifies the User's registration (authorization) and access (roles) to the requested resource identified by the `aud` claim.
 
 === An Example
 
-Say Bob has access to applications A, B and C configured in FusionAuth, and Lisa has access to applications C, D and E. Assume there's also an application F to which neither Bob nor Lisa is registered. 
+Say Bob has access to applications A, B and C configured in FusionAuth, and Lisa has access to applications C, D and E. Assume there's also an application F to which neither Bob nor Lisa is registered.  Note that for this example to work all the applications A thru F have "Require Registration" disabled (set to `false`).
```
Comment (brettwp) [inline site/docs/v1/tech/core-concepts/authentication-authorization.adoc:60]: @mooreds I added your suggestion above L52 and reiterated the idea here L60 to help enforce the idea of how "Require Registration" works.  Let me know what you think.

**Comment by robotdan** [comment]
> > Enable grants in response is missing `urn:ietf:params:oauth:grant-type:device_code`, see `GrantType`, this isn't related to your change, but while you're in here, can sync it up.
> 
> @robotdan This PR changes 3 `response.json` files and they all have `"enabledGrants": [...]` arrays. I'm not sure which ones to add `urn:ietf:params:oauth:grant-type:device_code` to or if that is even the correct place. Could you elaborate?

Sure, see `SupportedGrants` in `DefaultOAuthService` and the usage during validation. Each grant has a "grantName" property that is checked during validation to see if it is an allowed grant for this application. 

The allowed values should match this `SupportedGrants` value, which is:

```
authorization_code
implicit
password
refresh_token
urn:ietf:params:oauth:grant-type:device_code
```

**Comment by brettwp** [comment]
Okay, so just to clarify any where in "request body" or "response body" where we list the supported values for `application.oauthConfiguration.enabledGrants` I should update the list to be:
```
authorization_code
implicit
password
refresh_token
urn:ietf:params:oauth:grant-type:device_code
```
What about `client_credentials`?  I see that is in the `GrantType` enum, but not listed.

**Comment by mooreds** [comment]
@brettwp `client_credentials` grants don't take place with Applications, only with Entities. So you'd never have that value in  `application.oauthConfiguration.enabledGrants`.

**Comment by brettwp** [comment]
Thanks @mooreds.  That makes sense and I've updated the `device_code` entry.

File: site/docs/v1/tech/core-concepts/applications.adoc [hunk 7]
```diff
@@ -56,10 +56,10 @@ The tenant in which to create this Application.
 This field is only displayed once multiple tenants exist in FusionAuth. When only a single tenant exists, the Application will always be created in the default tenant.
 
 [field]#Theme# [optional]#Optional# [since]#Available since 1.27.0#::
-When a theme is selected, it will be used for this application instead of the tenant theme. 
+When a theme is selected, it will be used for this application instead of the tenant theme.
+
+A paid edition of FusionAuth is required to use the application themes.
```
Comment (robotdan) [inline site/docs/v1/tech/core-concepts/applications.adoc:61]: We could optionally update this to use the new "fancy" pattern for paid features.

```
+
:premium_feature: application themes
include::../shared/_premium-edition-blurb-api.adoc[]
+
:premium_feature!:
```

File: site/docs/v1/tech/core-concepts/applications.adoc [hunk 8]
```diff
@@ -152,6 +151,9 @@ The enabled OAuth2 grants. If a grant is not enabled and a client requests this
 +
 See The link:/docs/v1/tech/oauth/[OAuth 2.0 & OpenID Connect Overview] for additional information on each of these grants.
 
+[field]#Require Registration# [optional]#Optional# [default]#Defaults to `false`#::
```
Comment (robotdan) [inline site/docs/v1/tech/core-concepts/applications.adoc:154]: > Require registration

(lc registration)

File: site/docs/v1/tech/core-concepts/authentication-authorization.adoc (hunk 6 above)
Comment (robotdan) [reply site/docs/v1/tech/core-concepts/authentication-authorization.adoc:60]: > Require registration

(lc registration)

**Review by robotdan — COMMENTED** [review]
Added a few comments, otherwise LGTM.

File: site/docs/v1/tech/core-concepts/authentication-authorization.adoc (hunk 6 above)
Comment (brettwp) [reply site/docs/v1/tech/core-concepts/authentication-authorization.adoc]: Fixed

File: site/docs/v1/tech/core-concepts/applications.adoc (hunk 7 above)
Comment (brettwp) [reply site/docs/v1/tech/core-concepts/applications.adoc]: Changed.

**Pull request merged**